src/device_info.sh

```diff
@@ -0,0 +1,186 @@
+# This file deals with functions related to hardware information
+
+include "$KW_LIB_DIR/remote.sh"
+include "$KW_LIB_DIR/kwlib.sh"
+
+declare -gA shared_data=(['ram']='' ['cpu_model']='' ['cpu_currently']=''
+  ['cpu_max']='' ['cpu_min']='' ['disk_size']='' ['fs_name']='' ['fs_mount']=''
+  ['os']='')
+
+# This function populates the ram element from the shared_data global variable
+# with the total RAM memory from the target machine in kB
+#
+# @target Target machine
+function get_ram()
+{
+  local target="$1"
+  local ram
+  local cmd
+
+  cmd="cat /proc/meminfo | grep 'MemTotal' | grep -o '[0-9]*'"
+  case "$target" in
+    2) # LOCAL_TARGET
+      ram=$(cmd_manager 'SILENT' "$cmd")
+      ;;
+  esac
+
+  shared_data['ram']="$ram"
+}
+
+# This function provides the model and frequency of the CPU from a machine
+#
+# @target Target machine
+function get_cpu()
+{
+  local target="$1"
+  local cpu_model
+  local cpu_frequency
+  local cpu_currently
+  local cpu_max
+  local cpu_min
+
+  cmd_frequency="lscpu | grep MHz | sed -r 's/(CPU.*)/\t\t\1/'"
+  cmd_model="lscpu | grep 'Model name:' | sed -r 's/Model name:\s+//g' | cut -d' ' -f1"
+  case "$target" in
+    2) # LOCAL_TARGET
+      cpu_model=$(cmd_manager 'SILENT' "$cmd_model")
+      cpu_frequency=$(cmd_manager 'SILENT' "$cmd_frequency")
+      ;;
+  esac
+
+  cpu_currently=$(echo "$cpu_frequency" | grep 'CPU MHz')
+  cpu_max=$(echo "$cpu_frequency" | grep 'CPU max MHz')
+  cpu_min=$(echo "$cpu_frequency" | grep 'CPU min MHz')
+
+  cpu_currently=${cpu_currently//[!0-9,.]/}
+  cpu_max=${cpu_max//[!0-9,.]/}
+  cpu_min=${cpu_min//[!0-9,.]/}
+
+  shared_data['cpu_model']="$cpu_model"
+  shared_data['cpu_currently']="$cpu_currently"
+  shared_data['cpu_max']="$cpu_max"
+  shared_data['cpu_min']="$cpu_min"
+}
+
+# This function populates the values from the size and fs (filesystem) key of
+# the shared_data variable.
+#
+# @target Target machine
+function get_disk()
+{
+  local target="$1"
+  local cmd
+  local info
+  local size
+  local mount
+  local fs
+
+  cmd="df -h / | tail -n 1 | tr -s ' '"
+  case "$target" in
+    2) # LOCAL_TARGET
+      info=$(cmd_manager 'SILENT' "$cmd")
+      ;;
+  esac
+
+  fs="$(echo "$info" | cut -d' ' -f1)"
+  size="$(echo "$info" | cut -d' ' -f2)"
+  mount="$(echo "$info" | cut -d' ' -f6)"
+
+  shared_data['disk_size']="$size"
+  shared_data['fs_name']="$fs"
+  shared_data['fs_mount']="$mount"
+}
+
+# This function populates the os variable from the shared_data variable.
+#
+# @target Target machine
+function get_os()
+{
+  local target="$1"
+  local os
+
+  case "$target" in
+    2) # LOCAL_TARGET
+      os=$(detect_distro '/')
+      ;;
+  esac
+
+  shared_data['os']="$os"
+}
+
+# This function calls other functions to populate the shared_data variable with
+# the data related to the hardware from the target machine.
+#
+# @target Target machine
+function process_all_data()
```

Comment:
Inspired by ChromeOS scripts, let call this function `learn_device()` or similar.
See: https://chromium.googlesource.com/chromiumos/platform/crosutils/+/master/update_kernel.sh#67